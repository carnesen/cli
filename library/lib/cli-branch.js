"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CliBranch = void 0;
const constants_1 = require("./constants");
/**
 * A factory function for creating command branches
 * @remarks
 * @param options
 * @returns Returns the newly-created branch object
 * @example
 * ```typescript
 * const cloudBranch = CliBranch()
 * ```
 */
function CliBranch(options) {
    return {
        ...options,
        commandType: constants_1.CLI_BRANCH,
    };
}
exports.CliBranch = CliBranch;
//# sourceMappingURL=cli-branch.js.map