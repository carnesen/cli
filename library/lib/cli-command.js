"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CliCommand = void 0;
const constants_1 = require("./constants");
/**
 *
 * A factory function for creating CLI command commands
 * @param options
 * @returns The newly-created `command`.
 */
function CliCommand(options) {
    return {
        ...options,
        commandType: constants_1.CLI_COMMAND,
    };
}
exports.CliCommand = CliCommand;
//# sourceMappingURL=cli-command.js.map