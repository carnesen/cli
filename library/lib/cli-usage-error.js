"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CliUsageError = exports.CLI_USAGE_ERROR = void 0;
exports.CLI_USAGE_ERROR = 'CLI_USAGE_ERROR';
/**
 * A custom error with code {@linkcode CLI_USAGE_ERROR}
 * @remarks
 * {@linkcode runCliAndExit} catches this class of error and prints command-line usage
 * (i.e. "Usage: cloud users lists ...") for the current command. If the `message`
 * property is provided, `runCliAndExit` will normal `Error`. See also print `Error: <your
 * message>`. See also {@linkcode CliTerseError}.
 * @example
 * ```typescript
 * throw new CliUsageError("Expected one or more values")
 * ```
 */
class CliUsageError extends Error {
    /**
     *
     * @param message An `Error` message
     * @param commandStack Used internally for argument processing
     */
    constructor(message, commandStack) {
        super(message);
        Object.defineProperty(this, "code", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "commandStack", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.code = exports.CLI_USAGE_ERROR;
        this.commandStack = commandStack;
        Object.setPrototypeOf(this, new.target.prototype);
    }
}
exports.CliUsageError = CliUsageError;
//# sourceMappingURL=cli-usage-error.js.map