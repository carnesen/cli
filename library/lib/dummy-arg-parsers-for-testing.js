"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.dummyRequiredArgParser = exports.dummyArgParser = exports.DUMMY_ARG_PARSER_THROW_NON_TRUTHY = exports.DUMMY_ARG_PARSER_THROW = exports.DUMMY_ARG_PARSER_THROWN_INTENTIONALLY = exports.DUMMY_ARG_PARSER_EMPTY_ARRAY_WAS_PASSED = exports.DUMMY_ARG_PARSER_UNDEFINED_WAS_PASSED = void 0;
exports.DUMMY_ARG_PARSER_UNDEFINED_WAS_PASSED = 'undefined was passed';
exports.DUMMY_ARG_PARSER_EMPTY_ARRAY_WAS_PASSED = 'undefined was passed';
exports.DUMMY_ARG_PARSER_THROWN_INTENTIONALLY = 'thrown intentionally';
exports.DUMMY_ARG_PARSER_THROW = 'throw';
exports.DUMMY_ARG_PARSER_THROW_NON_TRUTHY = 'throw-non-truthy';
exports.dummyArgParser = {
    placeholder: '',
    parse(args) {
        if (typeof args === 'undefined') {
            return exports.DUMMY_ARG_PARSER_UNDEFINED_WAS_PASSED;
        }
        if (args.length === 0) {
            return exports.DUMMY_ARG_PARSER_EMPTY_ARRAY_WAS_PASSED;
        }
        if (args[0] === exports.DUMMY_ARG_PARSER_THROW) {
            throw new Error(exports.DUMMY_ARG_PARSER_THROWN_INTENTIONALLY);
        }
        if (args[0] === exports.DUMMY_ARG_PARSER_THROW_NON_TRUTHY) {
            // eslint-disable-next-line no-throw-literal
            throw '';
        }
        return args[0];
    },
};
exports.dummyRequiredArgParser = {
    placeholder: '<foo>',
    required: true,
    parse: exports.dummyArgParser.parse,
};
//# sourceMappingURL=dummy-arg-parsers-for-testing.js.map