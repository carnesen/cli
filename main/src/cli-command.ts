import { CLI_COMMAND } from './constants';
import {
  AnyArgParser,
  AnyNamedArgParsers,
  ValueFromArgParser,
  NamedValues,
  ICliArgParser,
} from './cli-arg-parser';

export type CliCommandOptions<
  TPositionalArgParser extends AnyArgParser,
  TNamedArgParsers extends AnyNamedArgParsers,
  TEscapedArgParser extends AnyArgParser
> = {
  /**
   * Name used in command-line usage
   */
  name: string;
  /**
   * The function that defines your command logic.
   *
   * @param positionalValue
   * @param namedValues
   * @param escapedValue
   * @returns A value (synchronously) or a `Promise` that resolves to a value
   * @remarks
   * By default, {@linkcode runCliAndExit} `console.log`s the value returned by `action`.
   * The type of the `args` argument received by `action` is derived by the `args`
   * property of the command. Similarly, the `options` argument type is derived from
   * `command.options`.
   *
   */
  action: (
    positionalValue: ValueFromArgParser<TPositionalArgParser>,
    namedValues: NamedValues<TNamedArgParsers>,
    escapedValue: ValueFromArgParser<TEscapedArgParser>,
  ) => any;
  positionalArgParser?: TPositionalArgParser;
  namedArgParsers?: TNamedArgParsers;
  escapedArgParser?: TEscapedArgParser;
  /**
   * A string that will be included in `Usage:` if present.
   */
  description?: string;
  /**
   * If true, don't show this command in usage docs.
   */
  hidden?: boolean;
};

/**
 * Interface describing a "command" command
 * @typeParam TPositionalArgParser Type of the "positional" arguments parser
 * @typeParam TNamedArgParsers Type of the ""
 */
export interface ICliCommand<
  TPositionalArgParser extends AnyArgParser,
  TNamedArgParsers extends AnyNamedArgParsers,
  TEscapedArgParser extends AnyArgParser
> extends CliCommandOptions<TPositionalArgParser, TNamedArgParsers, TEscapedArgParser> {
  commandType: typeof CLI_COMMAND;
}

/**
 *
 * A factory function for creating CLI command commands
 * @param options
 * @returns The newly-created `command`.
 */
export function CliCommand<
  TPositional extends AnyArgParser = ICliArgParser<undefined, false>,
  TNamed extends AnyNamedArgParsers = any,
  TEscaped extends AnyArgParser = ICliArgParser<undefined, false>
>(
  options: CliCommandOptions<TPositional, TNamed, TEscaped>,
): ICliCommand<TPositional, TNamed, TEscaped> {
  return {
    ...options,
    commandType: CLI_COMMAND,
  };
}
