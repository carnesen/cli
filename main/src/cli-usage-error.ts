import { Node } from './cli-node';

/** "code" of a {@link CliUsageError} */
export const CLI_USAGE_ERROR = 'CLI_USAGE_ERROR';

/**
 * Thrown to print command usage and an error message but not a `stack`
 */
export class CliUsageError extends Error {
  /** The string constant {@linkcode CLI_USAGE_ERROR} */
  public readonly code: typeof CLI_USAGE_ERROR;

  /** Used internally for constructing the command-line usage string */
  public node?: Node;

  /**
   * @param message If provided, {@linkcode runCliAndExit} will also print "Error: \<your
   * message\>"
   * @param node Used internally for constructing the command-line usage string
   */
  constructor(message?: string, node?: Node) {
    super(message);
    this.code = CLI_USAGE_ERROR;
    this.node = node;
    Object.setPrototypeOf(this, new.target.prototype);
  }
}
