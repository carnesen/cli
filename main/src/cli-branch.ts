import { ICliCommand } from './cli-command';

/** "kind" of an {@linkcode ICliBranch} */
export const CLI_BRANCH = 'CLI_BRANCH';

/** Options for {@linkcode CliBranch} */
export interface ICliBranchOptions {
  /** Name of this command branch, typically a category like "user" or "cloud" */
  name: string;

  /** A short description of this branch for command-line usage */
  description?: string;

  /** If `true`, this branch will not appear in command-line usage */
  hidden?: boolean;

  /** Branches and/or commands underneath this branch */
  children: (ICliBranch | ICliCommand<any, any, any>)[];
}

/**
 * A branch in a command tree
 */
export interface ICliBranch extends ICliBranchOptions {
  /** Used internally for discriminating between {@linkcode ICliBranch}'s and
   * {@linkcode ICliCommand}'s */
  kind: typeof CLI_BRANCH;
}

/**
 * A factory for {@linkcode ICliBranch} objects
 * @param options
 * @returns An {@linkcode ICliBranch} object
 * @example
 * ```typescript
 * const cloudBranch = CliBranch()
 * ```
 */
export function CliBranch(options: ICliBranchOptions): ICliBranch {
  return {
    ...options,
    kind: CLI_BRANCH,
  };
}
