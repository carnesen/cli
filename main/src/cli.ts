import { partitionArgs } from './partition-args';
import { findCliTree } from './find-cli-tree';
import { parseArgs } from './parse-args';
import { parseNamedArgs } from './parse-named-args';
import { CliUsageError, CLI_USAGE_ERROR } from './cli-usage-error';
import { TCliRoot } from './cli-tree';
import { CLI_COMMAND } from './cli-command';

/**
 * A command-line interface (CLI) function
 * */
export interface ICli {
	/**
	 * @param args The full set of command-line arguments. Defaults to
	 * `process.argv.slice(2)` in [[`runCli`]].
	 * @returns [Optional] The default CLI runner [[`runCli`]] will
	 * console.log the return value if there is one.
	 */
	(...args: string[]): Promise<any>;
}

/**
 * A factory for [[`ICli`]]s
 *
 * @param root The root of this command-line interface's command tree
 */
export function Cli(root: TCliRoot): ICli {
	return async function cli(...args: string[]) {
		const tree = findCliTree(root, args);

		if (tree.message || tree.current.kind !== CLI_COMMAND) {
			throw new CliUsageError(tree.message, tree);
		}

		const { positionalArgs, namedArgs, escapedArgs } = partitionArgs(tree.args);

		// We found "--help" among the arguments
		if (namedArgs.help) {
			throw new CliUsageError(undefined, tree);
		}

		const command = tree.current;

		// Pre-validation for positional argument group
		if (!command.positionalArgGroup && positionalArgs.length > 0) {
			throw new CliUsageError(
				`Unexpected argument "${positionalArgs[0]}" : Command "${command.name}" does not accept positional arguments`,
				tree,
			);
		}

		// All validation for named argument groups is done during parsing

		// Pre-validation for escaped argument group
		if (!command.escapedArgGroup && escapedArgs) {
			throw new CliUsageError(
				`Command "${command.name}" does not allow "--" as an argument`,
			);
		}

		// Calls to `parseArgs` and `action` in this try/catch block could throw
		// `CliUsageError`, which we catch and enhance with the current `TCliTree`
		// context.
		try {
			let argsValue: any;
			if (command.positionalArgGroup) {
				// Note that for named and escaped args, we distinguish between
				// `undefined` and `[]`. For example, "cli" gives an escaped args
				// `undefined` whereas "cli --" gives an escaped args `[]`. For the
				// "positionalArgs", however, there is no such distinction. By convention,
				// we elect here to pass in `undefined` rather than an empty array when no
				// positional arguments are passed.
				argsValue = await parseArgs(
					command.positionalArgGroup,
					positionalArgs.length > 0 ? positionalArgs : undefined,
					undefined,
				);
			}

			const namedValues = await parseNamedArgs(
				command.namedArgGroups || {},
				namedArgs,
			);

			let escapedValue: any;
			if (command.escapedArgGroup) {
				escapedValue = await parseArgs(
					command.escapedArgGroup,
					escapedArgs,
					'--',
				);
			}

			const result = await command.action(argsValue, namedValues, escapedValue);
			return result;
		} catch (exception) {
			if (exception && exception.code === CLI_USAGE_ERROR) {
				exception.tree = tree;
			}
			throw exception;
		}
	};
}
