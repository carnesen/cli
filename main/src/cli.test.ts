import { runAndCatch } from '@carnesen/run-and-catch';
import { CliBranch } from './cli-branch';
import { CliCommand } from './cli-command';
import { dummyValuedParser } from './dummy-arg-parsers-for-testing';
import { Cli, ICliEnhancer } from './cli';
import { CLI_USAGE_ERROR } from './cli-usage-error';

const commandWithNamedValuedParsers = CliCommand({
  name: 'command-with-named-args',
  namedParsers: {
    foo: dummyValuedParser,
  },
  action(...args) {
    return args;
  },
});

const commandWithPositionalValuedParser = CliCommand({
  name: 'command-with-positional-args',
  positionalParser: dummyValuedParser,
  action(...args) {
    return args;
  },
});

const commandWithEscapedValuedParser = CliCommand({
  name: 'command-with-escaped-arg-parser',
  escapedParser: dummyValuedParser,
  action(...args) {
    return args;
  },
});

const root = CliBranch({
  name: 'cli',
  children: [
    commandWithPositionalValuedParser,
    commandWithNamedValuedParsers,
    commandWithEscapedValuedParser,
  ],
});

const cliArgRunner = Cli(root);

describe(Cli.name, () => {
  it('calls the enhancer if provided', async () => {
    const spy = jest.fn();
    const enhancer: ICliEnhancer = (innerArgRunner) => async (...args: string[]) => {
      spy(...args);
      await innerArgRunner(...args);
    };
    const enhancedArgRunner = Cli(commandWithPositionalValuedParser, {
      enhancer,
    });
    await enhancedArgRunner('foo', 'bar');
    expect(spy.mock.calls).toEqual([['foo', 'bar']]);
  });

  it('throws USAGE error with empty message if --help is passed', async () => {
    const exception = await runAndCatch(cliArgRunner, '--help');
    expect(exception.code).toBe(CLI_USAGE_ERROR);
    expect(exception.message).toBeFalsy();
  });

  it('throws USAGE error with empty message if last command is a branch and no additional args is present', async () => {
    const exception = await runAndCatch(cliArgRunner);
    expect(exception.code).toBe(CLI_USAGE_ERROR);
    expect(exception.message).toBeFalsy();
  });

  it('throws USAGE error "bad command" if last command is a branch and additional args is present', async () => {
    const exception = await runAndCatch(cliArgRunner, 'oops');
    expect(exception.code).toBe(CLI_USAGE_ERROR);
    expect(exception.message).toMatch(/bad command/i);
    expect(exception.message).toMatch('"oops"');
    expect(exception.message).toMatchSnapshot();
  });

  it('throws USAGE error "positional arguments" if last command is a command without positionalValuedParser property and additional args is present', async () => {
    const exception = await runAndCatch(
      cliArgRunner,
      commandWithNamedValuedParsers.name,
      'oops',
    );
    expect(exception.code).toBe(CLI_USAGE_ERROR);
    expect(exception.message).toMatch('Unexpected argument "oops"');
    expect(exception.message).toMatch(commandWithNamedValuedParsers.name);
    expect(exception.message).toMatch(/positional arguments/i);
    expect(exception.message).toMatchSnapshot();
  });

  it('Passes parsed positional value as first argument of the "action" function', async () => {
    const positionalArgs = ['foo', 'bar'];
    const result = await cliArgRunner(
      commandWithPositionalValuedParser.name,
      ...positionalArgs,
    );
    expect(result).toEqual([dummyValuedParser.parse(positionalArgs), {}, undefined]);
  });

  it('Passes parsed named values as second argument of the "action" function', async () => {
    const namedArgs = ['--foo', 'bar'];
    const result = await cliArgRunner(commandWithNamedValuedParsers.name, ...namedArgs);
    expect(result).toEqual([
      undefined,
      { foo: dummyValuedParser.parse(['bar']) },
      undefined,
    ]);
  });

  it(`Throws USAGE error 'does not allow "--"' if command does not have an "escaped" property`, async () => {
    const exception = await runAndCatch(
      cliArgRunner,
      commandWithPositionalValuedParser.name,
      '--',
    );
    expect(exception.code).toBe(CLI_USAGE_ERROR);
    expect(exception.message).toMatch(commandWithPositionalValuedParser.name);
    expect(exception.message).toMatch('does not allow "--"');
  });

  it('Passes parsed escaped value as third argument of the "action" function', async () => {
    const result = await cliArgRunner(commandWithEscapedValuedParser.name, '--');
    expect(result).toEqual([
      undefined,
      {},
      commandWithEscapedValuedParser.escapedParser!.parse([]),
    ]);
  });
});
