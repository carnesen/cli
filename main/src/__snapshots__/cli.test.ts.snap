// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cli throws USAGE error "bad command" if last command is a branch and additional args is present 1`] = `"Bad command \\"oops\\""`;

exports[`Cli throws USAGE error "positional arguments" if last command is a command without positionalValuedParser property and additional args is present 1`] = `"Unexpected argument \\"oops\\" : Command \\"command-with-named-args\\" does not accept positional arguments"`;
